### PR TITLE
Minimize scratch disk initial size

### PR DESCRIPTION
**Pull request by barpavel**
Add to VDSM API and use 3 new properties that will allow creating scratch disk on block storage with small initial size.
VDSM will monitor the scratch disk block threshold and extend it as needed.

Signed-off-by: Pavel Bar <pbar@redhat.com>
Bug-Url: https://bugzilla.redhat.com/1913389

### CODE REVIEW

**Comment by bennyz** [comment]
why not build from source?
you can generate with `make rpm`

https://github.com/oVirt/vdsm#building-vdsm

File: lib/vdsm/virt/backup.py [hunk 1]
```diff
@@ -150,6 +150,9 @@ def __init__(self, backup_config):
 
 def start_backup(vm, dom, config):
     backup_cfg = BackupConfig(config)
+    log.info("BackupConfig object (s): %s", backup_cfg)
+    log.info("BackupConfig object (r): %r", backup_cfg)
```
Comment (nirs) [inline lib/vdsm/virt/backup.py:154]: All the info is already logged via the API layer. I guess that this is a debugging
log for development?

**Comment by nirs** [comment]
You don't need to add pull request for building the rpms. They are already built in
your own fork:
https://github.com/barpavel/vdsm/actions/runs/1813932423

Pull request is useful when you want to get a review - even if the work is not complete.

File: lib/vdsm/virt/backup.py (hunk 1 above)
Comment (barpavel) [reply lib/vdsm/virt/backup.py:154]: This is just to verify that the new VDSM version was installed.
These logs will help me to verify that :)

**Comment by barpavel** [comment]
`BackupConfig` object looks like this:
```
{
  "py/object": "vdsm.virt.backup.BackupConfig",
  "backup_id": "39fdae99-a2e4-427f-a7c6-3d01029fe3e6",
  "from_checkpoint_id": null,
  "to_checkpoint_id": "43222ab0-8dd6-4bb1-8b47-95bb32dccdb6",
  "require_consistency": false,
  "creation_time": null,
  "disks": [
    {
      "py/object": "vdsm.virt.backup.DiskConfig",
      "vol_id": "9db20b9d-e13f-44f6-8879-409526165dd6",
      "img_id": "f445641e-262c-42c1-a209-7c89dcd42058",
      "dom_id": "879c1040-dfd5-4f4e-b6db-e97890f6e0bb",
      "checkpoint": true,
      "backup_mode": "full",
      "scratch_disk": {
        "py/object": "vdsm.virt.backup.ScratchDiskConfig",
        "img_id": "81a39e1c-be03-47b5-81b0-52953818ba7c",
        "dom_id": "879c1040-dfd5-4f4e-b6db-e97890f6e0bb",
        "vol_id": "2283f8c4-c463-46e2-973c-5098e4505ef2",
        "path": "/rhev/data-center/mnt/10.35.0.164:_home_storage_nfs__domains_sd1/879c1040-dfd5-4f4e-b6db-e97890f6e0bb/images/81a39e1c-be03-47b5-81b0-52953818ba7c/2283f8c4-c463-46e2-973c-5098e4505ef2",
        "type": "file"
      }
    },
    {
      "py/object": "vdsm.virt.backup.DiskConfig",
      "vol_id": "4b7b9946-913a-4001-a5f6-58f544f26bb5",
      "img_id": "db97d6fd-ae3e-4e29-b045-bfd66e3bfae0",
      "dom_id": "879c1040-dfd5-4f4e-b6db-e97890f6e0bb",
      "checkpoint": true,
      "backup_mode": "full",
      "scratch_disk": {
        "py/object": "vdsm.virt.backup.ScratchDiskConfig",
        "img_id": "6affe083-4da8-45c4-83db-4fe9fc76432b",
        "dom_id": "879c1040-dfd5-4f4e-b6db-e97890f6e0bb",
        "vol_id": "780d2106-06d0-4f38-ae12-cd3d1e0b5f16",
        "path": "/rhev/data-center/mnt/10.35.0.164:_home_storage_nfs__domains_sd1/879c1040-dfd5-4f4e-b6db-e97890f6e0bb/images/6affe083-4da8-45c4-83db-4fe9fc76432b/780d2106-06d0-4f38-ae12-cd3d1e0b5f16",
        "type": "file"
      }
    }
  ]
}
```

File: lib/vdsm/virt/backup.py [hunk 2]
```diff
@@ -352,8 +352,20 @@ def _get_backup_disks(vm, backup_cfg):
```
Comment (nirs) [inline lib/vdsm/virt/backup.py:352]: @barpavel if this is a WIP, why did you add it to this PR, which should be ready for
review?

You can work on this in another branch. Please add to the PR when its ready
for review.

File: lib/vdsm/api/vdsm-api.yml [hunk 3]
```diff
@@ -8451,6 +8451,21 @@ types:
             name: path
```
Comment (nirs) [inline lib/vdsm/api/vdsm-api.yml:8451]: Notes about the commit message: I don't think we need these ordered lists, since these
items are not ordered. A simpler way to describe this change is:

    Add domainID, imageID, and volumeID to {File,Block}ScratchDisk.

This describe what this change does, but we need to understand why, and how the ids will be used.

File: lib/vdsm/api/vdsm-api.yml [hunk 4]
```diff
@@ -8451,6 +8451,21 @@ types:
...
+
+        -   description: The UUID of the Storage Domain containing the Volume
+            name: domainID
+            type: *UUID
+            added: '4.5'
+
+        -   description: The UUID of the Volume
+            name: volumeID
+            type: *UUID
+            added: '4.5'
```
Comment (nirs) [inline lib/vdsm/api/vdsm-api.yml:8467]: Please check the order of the properties in BackupDisk:

        -   description: The UUID of the storage domain
            name: sd_id
            type: *UUID
        -   description: The UUID of the image
            name: img_id
            type: *UUID
        -   description: The UUID of the volume
            name: vol_id
            type: *UUID

This order is not accidental - we have a name for these ids - `DIV` - domain, image, volume. We want to keep this order everywhere.

Please update the *ScatchDisk types to have the standard order.

Additionally, we are missing a fix for BackupDisk keys - the actual API (what engine
is sending) is domainID, imageID, volumeID, so we should replace sd_id, img_id, vol_id 
with the actual names.

This can be done in another PR, and opening new issue for this will be good so we
don't forget to do this.

File: lib/vdsm/virt/backup.py [hunk 5]
```diff
@@ -74,6 +74,9 @@ class ScratchDiskConfig(properties.Owner):
         values=[DISK_TYPE.FILE, DISK_TYPE.BLOCK])
 
     def __init__(self, **kw):
+        self.img_id = kw.get("img_id")
+        self.dom_id = kw.get("dom_id")
+        self.vol_id = kw.get("vol_id")
         self.path = kw.get("path")
         self.type = kw.get("type")
```
Comment (nirs) [inline lib/vdsm/virt/backup.py:81]: Please add the keys in the same order they appear in the schema. We want to the
to have the same structure in the schema, in vdsm internal implementation, and if
possible in on engine side.

Please add properties to the class - see how path and type are added. The properties provide validation for the values. See how we define the same keys for BackupDisk.

Note that the ids are not required, since when using transient disk, we don't have
any id.

File: lib/vdsm/virt/backup.py [hunk 6]
```diff
@@ -97,6 +100,9 @@ def __init__(self, disk_config):
         if "scratch_disk" in disk_config:
             scratch_disk = disk_config.get("scratch_disk")
             self.scratch_disk = ScratchDiskConfig(
+                img_id=scratch_disk.get("imageID"),
+                dom_id=scratch_disk.get("domainID"),
+                vol_id=scratch_disk.get("volumeID"),
                 path=scratch_disk.get("path"),
                 type=scratch_disk.get("type"))
```
Comment (nirs) [inline lib/vdsm/virt/backup.py:107]: Please update after changing the order of the instance variable.

File: lib/vdsm/virt/backup.py [hunk 7]
```diff
@@ -667,7 +673,13 @@ def _create_scratch_disks(vm, dom, backup_id, backup_disks):
         except Exception:
             _remove_scratch_disks(vm, backup_id)
             raise
-        backup_disk.scratch_disk = ScratchDiskConfig(path=path, type="file")
+        backup_disk.scratch_disk = ScratchDiskConfig(
+            # TODO: what should be the values of these 3 fields?
+            img_id=None,
+            dom_id=None,
+            vol_id=None,
```
Comment (nirs) [inline lib/vdsm/virt/backup.py:700]: Since this scratch disk is not a transient disk and not an image, they don't have 
the ids. No change is needed in this code - if you do't specify the ids, using;

    self.name = kw.get("name")

Will initialize the instance variables to None.

Not having ids is not an issue for the code adding the scratch disk ids
to the the scratch_disk dict in the drive, since we don't keep file based
scratch disks.

File: lib/vdsm/virt/backup.py [hunk 8]
```diff
@@ -365,6 +377,14 @@ def _start_monitoring_scratch_disks(vm, backup_disks, backup):
             }
```
Comment (nirs) [inline lib/vdsm/virt/backup.py:377]: I don't know why you need so much debug prints, when the actual change is just:

    "dom_id": backup_disk.scratch_disk.dom_id,
    ...

And updating the to print backup_disk.drive.scratch_disk instead of formatting all
the info in the log.  The log will show the scratch disk in a very clear way, there is no need for all this debug effort.

**Review by nirs — CHANGES_REQUESTED** [review]
Thanks @barpavel!

**Comment by nirs** [comment]
> `BackupConfig` object looks like this:
> 
> ```
> {
>   "py/object": "vdsm.virt.backup.BackupConfig",

This is not part of the BackupConfig object but artifact of the tool you used
to create this json (jsonpickle?).

...
>       "scratch_disk": {
>         "py/object": "vdsm.virt.backup.ScratchDiskConfig",
>         "img_id": "81a39e1c-be03-47b5-81b0-52953818ba7c",
>         "dom_id": "879c1040-dfd5-4f4e-b6db-e97890f6e0bb",
>         "vol_id": "2283f8c4-c463-46e2-973c-5098e4505ef2",

Looks right and matches the path.

>         "path": "/rhev/data-center/mnt/10.35.0.164:_home_storage_nfs__domains_sd1/879c1040-dfd5-4f4e-b6db-e97890f6e0bb/images/81a39e1c-be03-47b5-81b0-52953818ba7c/2283f8c4-c463-46e2-973c-5098e4505ef2",
>         "type": "file"
>       }
```

**Comment by barpavel** [comment]
> > `BackupConfig` object looks like this:
> > ```
> > {
> >   "py/object": "vdsm.virt.backup.BackupConfig",
> > ```
> 
> This is not part of the BackupConfig object but artifact of the tool you used to create this json (jsonpickle?).
> 
> ...
> 
> > ```
> >   "scratch_disk": {
> >     "py/object": "vdsm.virt.backup.ScratchDiskConfig",
> >     "img_id": "81a39e1c-be03-47b5-81b0-52953818ba7c",
> >     "dom_id": "879c1040-dfd5-4f4e-b6db-e97890f6e0bb",
> >     "vol_id": "2283f8c4-c463-46e2-973c-5098e4505ef2",
> > ```
> 
> Looks right and matches the path.
> 
> > ```
> >     "path": "/rhev/data-center/mnt/10.35.0.164:_home_storage_nfs__domains_sd1/879c1040-dfd5-4f4e-b6db-e97890f6e0bb/images/81a39e1c-be03-47b5-81b0-52953818ba7c/2283f8c4-c463-46e2-973c-5098e4505ef2",
> >     "type": "file"
> >   }
> > ```

Updated the JSON above to not have the `py/object` and also to have the `sd_id` instead of `dom_id`.

**Comment by nirs** [comment]
@barpavel the tests fail - you need to fix this.

File: lib/vdsm/virt/backup.py [hunk 9]
```diff
@@ -355,13 +355,14 @@ def _start_monitoring_scratch_disks(vm, backup_disks, backup):
...
-            # TODO: Add domain, image, and volume ids when they are
-            # availble.
             scratch_disk = disks[backup_disk.drive.name]
             vm.log.info("Start monitoring scratch disk %s for drive %s",
                         scratch_disk, backup_disk.drive.name)
             backup_disk.drive.scratch_disk = {
                 "index": scratch_disk["index"],
+                "sd_id": backup_disk.scratch_disk.sd_id,
+                "img_id": backup_disk.scratch_disk.img_id,
+                "vol_id": backup_disk.scratch_disk.vol_id,
```
Comment (nirs) [inline lib/vdsm/virt/backup.py:365]: Change looks good now, but it will add:

    "sd_id": None

When engine did not send the uuids. In this case we cannot monitor the scratch disk
so we don't need the ids in the dict.

Lets add the ids only of the they are not None. For example:

    if backup_disk.scratch_disk.sd_id is not None:
        backup_disk.drive.scratch_disk["sd_id"] = backup_disk.scratch_disk.sd_id

This also probably breaks backup tests, checking for the old
content of the dict.

I guess you need to update the tests - we should test:
- old way: engine sends no scratch disk ids
- new way: engine sends scratch disks ids

You can run the backup tests locally with:

    tox -e virt tests/virt/backup_test.py

The tests run very quickly, no need to push changes to github and wait 5 minutes until
the all the tests pass.

**Review by nirs — CHANGES_REQUESTED** [review]

File: lib/vdsm/virt/backup.py (hunk 5 above)
Comment (barpavel) [reply lib/vdsm/virt/backup.py]: Actually for `BackupDisk` I don't see any properties defined, unless I'm missing something :)
But I believe I understood what you suggested:
Added the 3 properties:
```
    sd_id = properties.UUID(required=False)
    img_id = properties.UUID(required=False)
    vol_id = properties.UUID(required=False)
```

**Comment by barpavel** [comment]
> Thanks @barpavel!

Welcome :)

**Comment by barpavel** [comment]
> @barpavel the tests fail - you need to fix this.

Done.
Your suggestion to check for `None` before adding to the `scratch_disk` dictionary not only eliminated redundant mappings from `sd_id`, `img_id` & `vol_id` to `None`, but also solved the tests issue.

**Review by nirs — CHANGES_REQUESTED** [review]

**Comment by barpavel** [comment]
> > > @barpavel the tests fail - you need to fix this.
> > 
> > 
> > Done. Your suggestion to check for `None` before adding to the `scratch_disk` dictionary not only eliminated redundant mappings from `sd_id`, `img_id` & `vol_id` to `None`, but also solved the tests issue.
> 
> It solved the issue because we don't have a test creating backup with ids. You need to add such test.

Will do.

File: lib/vdsm/virt/backup.py [hunk 10]
```diff
@@ -358,14 +358,21 @@ def _start_monitoring_scratch_disks(vm, backup_disks, backup):
...
-            # TODO: Add domain, image, and volume ids when they are
-            # availble.
             scratch_disk = disks[backup_disk.drive.name]
             vm.log.info("Start monitoring scratch disk %s for drive %s",
                         scratch_disk, backup_disk.drive.name)
             backup_disk.drive.scratch_disk = {
                 "index": scratch_disk["index"],
             }
+            if backup_disk.scratch_disk.sd_id is not None:
+                backup_disk.drive.scratch_disk["sd_id"] =\
```
[the remark this answers is not in the document]
Comment (barpavel) [reply lib/vdsm/virt/backup.py]: Done.

**Comment by barpavel** [comment]
> > > > @barpavel the tests fail - you need to fix this.
> > > 
> > > 
> > > Done. Your suggestion to check for `None` before adding to the `scratch_disk` dictionary not only eliminated redundant mappings from `sd_id`, `img_id` & `vol_id` to `None`, but also solved the tests issue.
> > 
> > 
> > It solved the issue because we don't have a test creating backup with ids. You need to add such test.
> 
> Will do.

Tried to add it, but it got complicated, fails on `assert normalized(dom.backupGetXMLDesc()) == normalized(backup_xml)`. Need to understand where the 1st XML comes from.

File: lib/vdsm/virt/backup.py (hunk 9 above)
Comment (nirs) [reply lib/vdsm/virt/backup.py]: > Fixed according to your suggestion. While looking for a more elegant solution encountered the following arguably better implementation :)
> 
> ```
> {k: v for k, v in (('orange', orange), ('apple', apple)) if v is not None}
> ```

This works well for converting list of tuples to dict, removing None values. But we
don't have such list, and the simple approach is simpler.

File: tests/virt/backup_test.py [hunk 11]
```diff
@@ -362,6 +363,90 @@ def test_start_stop_backup_engine_scratch_disks(tmpdir):
...
+    socket = backup.socket_path(BACKUP_1_ID)
+    scratch1 = create_scratch_disk(tmpdir, "scratch1")
+    scratch2 = create_scratch_disk(tmpdir, "scratch2")
+
+    fake_disks = create_fake_disks(vm)
+
+    # Engine sends scratch disk IDs.
+    scratch_disk_domain_uuid = make_uuid()
+    scratch_disk_image_uuid = make_uuid()
+    scratch_disk_volume_uuid = make_uuid()
```
Comment (nirs) [inline tests/virt/backup_test.py:379]: These variables are not needed

File: tests/virt/backup_test.py [hunk 12]
```diff
@@ -362,6 +363,90 @@ def test_start_stop_backup_engine_scratch_disks(tmpdir):
...
+
+    # Block based scratch disk for block based disk.
+    fake_disks[0]['scratch_disk'] = {
+        'path': scratch1,
+        'type': DISK_TYPE.BLOCK,
+        'domainID': scratch_disk_domain_uuid,
```
Comment (nirs) [inline tests/virt/backup_test.py:385]: You can use make_uuid() to create new uuid

File: tests/virt/backup_test.py [hunk 13]
```diff
@@ -362,6 +363,90 @@ def test_start_stop_backup_engine_scratch_disks(tmpdir):
...
+        'volumeID': scratch_disk_volume_uuid,
+    }
+
+    # File based scratch disk for file based disk.
+    fake_disks[1]['scratch_disk'] = {
+        'path': scratch2,
+        'type': DISK_TYPE.FILE,
+        'domainID': scratch_disk_domain_uuid,
+        'imageID': scratch_disk_image_uuid,
+        'volumeID': scratch_disk_volume_uuid,
```
Comment (nirs) [inline tests/virt/backup_test.py:396]: Using same uuid for block and file volume is wrong. Using make_uuir() will solve this.

File: tests/virt/backup_test.py [hunk 14]
```diff
@@ -362,6 +363,90 @@ def test_start_stop_backup_engine_scratch_disks(tmpdir):
...
+        </domainbackup>
+        """
+    assert normalized(dom.backupGetXMLDesc()) == normalized(backup_xml)
+
+    # We monitor only block based scratch disks.
+    assert vm.drives[IMAGE_1_UUID].scratch_disk == {
+        "index": 7,
+        'sd_id': scratch_disk_domain_uuid,
+        'img_id': scratch_disk_image_uuid,
+        'vol_id': scratch_disk_volume_uuid,
```
Comment (nirs) [inline tests/virt/backup_test.py:435]: you can get the right uuids from:

    fake_disks[0]['scratch_disk']['domainID']

**Review by nirs — CHANGES_REQUESTED** [review]
New test looks good, needs minor cleanup.

Did you verify that the new test fail when we don't add the uuids to the scratch_disk dict?

**Comment by barpavel** [comment]
> @barpavel let’s make small steps. If current pr works and tested, we can merge it now and add the missing test in in the next pr.

@nirs Fixed the tests. Please check it :)

**Comment by barpavel** [comment]
> > @barpavel let’s make small steps. If current pr works and tested, we can merge it now and add the missing test in in the next pr.
> 
> @nirs Fixed the tests. Please check it :)

@nirs Sorry, just noticed that you already left a few comments today morning, I didn't fix the tests according to you today's comments, doing it now.

File: tests/virt/backup_test.py (hunk 11 above)
Comment (barpavel) [reply tests/virt/backup_test.py]: Done.
Inlined the 3 variables.

File: tests/virt/backup_test.py (hunk 12 above)
Comment (barpavel) [reply tests/virt/backup_test.py]: Done.
Inlined the 3 variables.

File: tests/virt/backup_test.py (hunk 13 above)
Comment (barpavel) [reply tests/virt/backup_test.py]: Done.

File: tests/virt/backup_test.py (hunk 14 above)
Comment (barpavel) [reply tests/virt/backup_test.py]: Done.

**Comment by barpavel** [comment]
> Did you verify that the new test fail when we don't add the uuids to the scratch_disk dict?

I didn't quite understand what do you mean :(
It is legal for engine to not send the scratch disks IDs, there is an old test for that, that I just renamed from `test_start_stop_backup_engine_scratch_disks` to `test_start_stop_backup_engine_scratch_disks_without_scratch_ids`.

**Comment by barpavel** [comment]
> > > @barpavel let’s make small steps. If current pr works and tested, we can merge it now and add the missing test in in the next pr.
> > 
> > 
> > @nirs Fixed the tests. Please check it :)
> 
> @nirs Sorry, just noticed that you already left a few comments today morning, I didn't fix the tests according to you today's comments, doing it now.

All comments fixed, except 1 question of yours: `Did you verify that the new test fail when we don't add the uuids to the scratch_disk dict?`

**Comment by barpavel** [comment]
> New test looks good, needs minor cleanup.
> 
> Did you verify that the new test fail when we don't add the uuids to the scratch_disk dict?

Yes, the tests fail if we don't add the scratch disks:
```
=================================== FAILURES ===================================
_________ test_start_stop_backup_engine_scratch_disks_with_scratch_ids _________

tmpdir = local('/var/tmp/vdsm/test_start_stop_backup_engine_1')

    def test_start_stop_backup_engine_scratch_disks_with_scratch_ids(tmpdir):

        ............ Many more lines ......

        assert normalized(dom.backupGetXMLDesc()) == normalized(backup_xml)
    
        # We monitor only block based scratch disks.
>       assert vm.drives[IMAGE_1_UUID].scratch_disk == {
            "index": 7,
            'sd_id': fake_disks[0]['scratch_disk']['domainID'],
            'img_id': fake_disks[0]['scratch_disk']['imageID'],
            'vol_id': fake_disks[0]['scratch_disk']['volumeID'],
        }
E       AssertionError: assert {'index': 7} == {'img_id': 'e...4c625b64dfd8'}
E         Omitting 1 identical items, use -vv to show
E         Right contains 3 more items:
E         {'img_id': 'e2b4315f-dd9a-4008-a5b6-1fd7f19ee46b',
E          'sd_id': '49ab81be-042c-49ac-9234-5a3d64afcb98',
E          'vol_id': '79c991d7-15e4-419a-b431-4c625b64dfd8'}
E         Full diff:
E           {
E         -  'img_id': 'e2b4315f-dd9a-4008-a5b6-1fd7f19ee46b',
E            'index': 7,
E         -  'sd_id': '49ab81be-042c-49ac-9234-5a3d64afcb98',
E         -  'vol_id': '79c991d7-15e4-419a-b431-4c625b64dfd8',
E           }

virt/backup_test.py:429: AssertionError
```

**Review by nirs — DISMISSED** [review]

**Comment by nirs** [comment]
@barpavel can you describe how this was tested?
- did you test this with old engine that does not send the uuids?

**Comment by barpavel** [comment]
> @barpavel can you describe how this was tested?
> 
> * did you test this with old engine that does not send the uuids?

Tested now, everything works just fine, VDSM handles correctly backup configuration without scratch disks' IDs.
To be sure I checked VDSM logs and verified that the backup configuration as received from the engine looks like that:
```
{
  "backup_id": "08d78cc2-5b8d-4609-98fd-4ed527351a48",
  "disks": [
    {
      "checkpoint": true,
      "scratch_disk": {
        "path": "/rhev/data-center/mnt/10.35.0.164:_home_storage_nfs__domains_sd1/4df89efd-61ed-4417-82cc-053ba98b6c7b/images/bf2e9a78-84e1-4b65-8d8b-3e80a50fe0ec/a6758a78-6f8f-4482-917c-8563300b2811",
        "type": "file"
      },
      "imageID": "4f5dbc5e-f61b-4768-a05c-f32e26a546e9",
      "volumeID": "abc19882-84f6-47d7-8b76-5542e8a77581",
      "backup_mode": "full",
      "domainID": "4df89efd-61ed-4417-82cc-053ba98b6c7b"
    },
    {
      "checkpoint": true,
      "scratch_disk": {
        "path": "/rhev/data-center/mnt/10.35.0.164:_home_storage_nfs__domains_sd1/4df89efd-61ed-4417-82cc-053ba98b6c7b/images/13aca210-55ca-4fc4-aaa1-d22b6a7c28dc/3dc88358-03fa-4927-81ec-e72f4e886f08",
        "type": "file"
      },
      "imageID": "37cc9514-3443-4ad9-abce-e7f5eb96eb7a",
      "volumeID": "827a0035-4c74-4d3e-8ed0-049fe735e4d0",
      "backup_mode": "full",
      "domainID": "4df89efd-61ed-4417-82cc-053ba98b6c7b"
    }
  ],
  "from_checkpoint_id": null,
  "parent_checkpoint_id": null,
  "require_consistency": false,
  "to_checkpoint_id": "50546bce-b14b-42be-979b-d65da6d505b5"
}
```

And as a result the scratch disk dictionary as created by `_start_monitoring_scratch_disks` is a simple:
```
{
  "index": 5
}
```

**Comment by nirs** [comment]
Thanks @barpavel, I think its time to merge this.

**Review by nirs — APPROVED** [review]
